### PR TITLE
An attemp to fix the text selection bug

### LINKED ISSUE #1641: When entering text, selecting all, and then using Command+Z to undo results in an error.
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/singerdmx/flutter-quill/issues)

### Flutter Quill version

9.0+

### Other Flutter Quill packages versions

_No response_

### Steps to reproduce

When entering text, selecting all, and then using Command+Z.

### Expected results

undo

### Actual results

error

### Code sample

<details><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

 
Xnip2023-12-22_11-06-57

### Logs

<details><summary>Logs</summary>

```console
Restarted application in 584ms.

════════ Exception caught by services library ══════════════════════════════════
The following assertion was thrown while processing the key message handler:
'dart:ui/text.dart': Failed assertion: line 2520 pos 16: '<optimized out>': is not true.

When the exception was thrown, this was the stack:
#2      new TextRange.collapsed (dart:ui/text.dart:2520:16)
#3      new TextSelection.collapsed (package:flutter/src/services/text_editing.dart:36:14)
#4      QuillController._handleHistoryChange (package:flutter_quill/src/widgets/quill/quill_controller.dart:254:25)
#5      QuillController.undo (package:flutter_quill/src/widgets/quill/quill_controller.dart:243:7)
#6      QuillEditorUndoKeyboardAction.invoke (package:flutter_quill/src/widgets/raw_editor/raw_editor_actions.dart:364:24)
#7      _OverridableContextAction.invokeDefaultAction (package:flutter/src/widgets/actions.dart:1811:28)
#8      _OverridableActionMixin.invoke (package:flutter/src/widgets/actions.dart:1676:9)
#9      Action._invoke (package:flutter/src/widgets/actions.dart:336:19)
#10     ActionDispatcher.invokeActionIfEnabled (package:flutter/src/widgets/actions.dart:666:28)
#11     ShortcutManager.handleKeypress (package:flutter/src/widgets/shortcuts.dart:851:83)
#12     _ShortcutsState._handleOnKey (package:flutter/src/widgets/shortcuts.dart:1042:20)
#13     _HighlightModeManager.handleKeyMessage (package:flutter/src/widgets/focus_manager.dart:1837:32)
#14     KeyEventManager._dispatchKeyMessage (package:flutter/src/services/hardware_keyboard.dart:996:34)
#15     KeyEventManager.handleRawKeyMessage (package:flutter/src/services/hardware_keyboard.dart:1061:17)
#16     BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:223:49)
#17     _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:567:35)
#18     _invoke2 (dart:ui/hooks.dart:344:13)
#19     _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:45:5)
#20     _Channel.push (dart:ui/channel_buffers.dart:135:31)
#21     ChannelBuffers.push (dart:ui/channel_buffers.dart:343:17)
#22     PlatformDispatcher._dispatchPlatformMessage (dart:ui/platform_dispatcher.dart:722:22)
#23     _dispatchPlatformMessage (dart:ui/hooks.dart:257:31)
(elided 2 frames from class _AssertionError)

KeyMessage: KeyMessage([KeyDownEvent#c441f(physicalKey: PhysicalKeyboardKey#c222c(usbHidUsage: "0x0007001d", debugName: "Key Z"), logicalKey: LogicalKeyboardKey#5989a(keyId: "0x0000007a", keyLabel: "Z", debugName: "Key Z"), character: "z", timeStamp: 19:19:03.247687)])
════════════════════════════════════════════════════════════════════════════════

```

</details>

### PR DESCRIPTION
**Pull request by ellet0**
- An attempt to fix #1641

### CODE REVIEW

**Review by singerdmx — APPROVED** [review]

**Review by singerdmx — APPROVED** [review]

**Comment by T-P-F** [comment]
@singerdmx Thanks for Flutter Quill. 

Curious when will this fix be merged? We're in need of it as well.

**Comment by ellet0** [comment]
> @singerdmx Thanks for Flutter Quill. 
> 
> Curious when will this fix be merged? We're in need of it as well.

I will review it once again today and then merge it

File: lib/src/widgets/quill/quill_controller.dart [hunk 1]
```diff
@@ -251,8 +251,13 @@ class QuillController extends ChangeNotifier {
       // updateSelection(
       // TextSelection.collapsed(offset: document.length), ChangeSource.LOCAL);
       updateSelection(
-          TextSelection.collapsed(offset: selection.baseOffset + len),
-          ChangeSource.local);
+        (selection.baseOffset + len) > 0
+            ? TextSelection.collapsed(
+                offset: selection.baseOffset + len,
```
Comment (ellet0) [inline lib/src/widgets/quill/quill_controller.dart:256]: Hi @adil192 
Do you know what is len?? in the class HistoryChanged?

File: lib/src/widgets/quill/C
Comment (EgorK0rshun) [inline lib/src/widgets/quill/C:1]: unnecessary file

Comment (ellet0) [reply lib/src/widgets/quill/C:1]: > unnecessary file

Thanks! Don't remember how I created it, I use the command line a lot and sometimes I can make such typos

File: lib/src/widgets/quill/quill_controller.dart (hunk 1 above)
Comment (adil192) [reply lib/src/widgets/quill/quill_controller.dart:256]: https://github.com/singerdmx/flutter-quill/pull/1128#issuecomment-1904276303

File: lib/src/models/documents/document.dart [hunk 2]
```diff
@@ -218,7 +218,7 @@ class Document {
     if (res.node is Line) {
       return res;
     }
-    final block = res.node as Block;
+    final block = res.node as Block; // TODO: Can be nullable, handle this case
```
Comment (adil192) [inline lib/src/models/documents/document.dart:221]: If this can be nullable, shouldn't it be `as Block?`

**Pull request merged**